google-cloud-storage/src/main/java/com/google/cloud/storage/RestartableInputStream.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.storage;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ *
+ */
+public class RestartableInputStream {
+
+    InputStream inputStream;
+
+    public RestartableInputStream(InputStream inputStream) {
+        this.inputStream = inputStream;
+        this.inputStream.mark(Integer.MAX_VALUE);
+    }
+
+    public InputStream getInputStream() {
+        return this.inputStream;
+    }
+
+    public void restart() {
+        try {
+            this.inputStream.reset();
+            this.inputStream.mark(Integer.MAX_VALUE);
```

Comment:
`Integer.MAX_VALUE` here and in the constructor seems too generous. It depends on the implementation, but there is a good chance, that the `InputStream`, which supports `mark/reset` does it by buffering its content in memory. Calling `mark(Integer.MAX_VALUE)` on it would cause it to buffer the whole content of the stream each time (even for the cases, when there were no retry needed, which should be 99% of the cases at least). For big streams (gigabytes+) this would consume too much resources and can easily throw `OutOfMemoryException`. We need to find an acceptable way to guarantee that it will not happen.

Reply:
In this case, how about taking the idea from `PushbackInputStream` but instead caching the bytes in memory, creating a new `FileOutputStream` to save the cached bytes to file system, and when retrying, creating a new `FileInputStream` to read back those bytes?